rust/src/encodings/binary.rs

```diff
@@ -1,3 +1,96 @@
-//! Binary encoding
+//! Var-length Binary Encoding
 //!
+
+use std::io::Result;
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+use arrow_array::types::{ByteArrayType, Int64Type};
+use arrow_array::{Array, ArrayRef, GenericByteArray, Int32Array, Int64Array};
+use arrow_data::ArrayDataBuilder;
+use async_trait::async_trait;
+
+use crate::encodings::Decoder;
+use crate::io::object_reader::ObjectReader;
+
+use super::plain::PlainDecoder;
+
 pub struct BinaryEncoder {}
+
+/// Var-binary encoding decoder.
+pub struct BinaryDecoder<'a, T: ByteArrayType> {
+    reader: &'a ObjectReader<'a>,
+
+    position: usize,
+
+    length: usize,
+
+    phantom: PhantomData<T>,
+}
+
+/// Var-length Binary Decoder
+///
+impl<'a, T: ByteArrayType> BinaryDecoder<'a, T> {
+    /// Create a [BinaryEncoder] to decode one batch.
+    ///
+    ///  - `position`, file position where this batch starts.
+    ///  - `length`, the number of records in this batch.
+    ///
+    /// ## Example
+    ///
+    /// ```rust
+    /// use arrow_array::types::Utf8Type;
+    /// use object_store::path::Path;
+    /// use lance::io::ObjectStore;
+    /// use lance::encodings::binary::BinaryDecoder;
+    ///
+    /// async {
+    ///     let object_store = ObjectStore::new(":memory:").unwrap();
+    ///     let path = Path::from("/data.lance");
+    ///     let reader = object_store.open(&path).await.unwrap();
+    ///     let string_decoder = BinaryDecoder::<Utf8Type>::new(&reader, 100, 1024);
+    /// };
+    /// ```
+    pub fn new(reader: &'a ObjectReader, position: usize, length: usize) -> Self {
+        Self {
+            reader,
+            position,
+            length,
+            phantom: PhantomData,
+        }
+    }
+}
+
+#[async_trait]
+impl<'a, T: ByteArrayType> Decoder for BinaryDecoder<'a, T> {
+    async fn decode(&self) -> Result<ArrayRef> {
+        let position_decoder =
+            PlainDecoder::<Int64Type>::new(self.reader, self.position, self.length + 1)?;
```

Comment:
not a big deal but should positions be Int64 or UInt64? does it matter at all?

Reply:
This is how the C++ side is serialized, it does not matter that much. Changing to u64 will be a break change, tho.